browser/browsing_data/brave_clear_browsing_data.cc

```diff
@@ -0,0 +1,182 @@
+
+  // BrowsingDataRemover::Observer implementation.
+  void OnBrowsingDataRemoverDone() override;
+
+ private:
+  bool GetClearBrowsingDataOnExitSettings(const Profile* profile,
+                                          int* remove_mask,
+                                          int* origin_mask);
+  void Wait();
+
+  int num_profiles_to_clear_ = 0;
+  base::RunLoop run_loop_;
+  // Keep track of the set of BrowsingDataRemover instances this object has
+  // attached itself to as an observer. When ScopedObserver is destroyed it
+  // removes this object as an observer from all those instances.
+  ScopedObserver<content::BrowsingDataRemover,
+                 content::BrowsingDataRemover::Observer>
+      observer_;
+};
+
+// See ClearBrowsingDataHandler::HandleClearBrowsingData which constructs the
+// remove_mask and the origin_mask for the same functionality not on exit.
+bool BrowsingDataRemovalWatcher::GetClearBrowsingDataOnExitSettings(
+    const Profile* profile,
+    int* remove_mask,
+    int* origin_mask) {
+  DCHECK(remove_mask);
+  DCHECK(origin_mask);
+  const PrefService* prefs = profile->GetPrefs();
+  *remove_mask = 0;
+  *origin_mask = 0;
+
+  int site_data_mask = ChromeBrowsingDataRemoverDelegate::DATA_TYPE_SITE_DATA;
+  // Don't try to clear LSO data if it's not supported.
+  if (!prefs->GetBoolean(prefs::kClearPluginLSODataEnabled))
+    site_data_mask &= ~ChromeBrowsingDataRemoverDelegate::DATA_TYPE_PLUGIN_DATA;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteBrowsingHistoryOnExit) &&
+      prefs->GetBoolean(prefs::kAllowDeletingBrowserHistory))
+    *remove_mask |= ChromeBrowsingDataRemoverDelegate::DATA_TYPE_HISTORY;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteDownloadHistoryOnExit) &&
+      prefs->GetBoolean(prefs::kAllowDeletingBrowserHistory))
+    *remove_mask |= content::BrowsingDataRemover::DATA_TYPE_DOWNLOADS;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteCacheOnExit))
+    *remove_mask |= content::BrowsingDataRemover::DATA_TYPE_CACHE;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteCookiesOnExit)) {
+    *remove_mask |= site_data_mask;
+    *origin_mask |= content::BrowsingDataRemover::ORIGIN_TYPE_UNPROTECTED_WEB;
+  }
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeletePasswordsOnExit))
+    *remove_mask |= ChromeBrowsingDataRemoverDelegate::DATA_TYPE_PASSWORDS;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteFormDataOnExit))
+    *remove_mask |= ChromeBrowsingDataRemoverDelegate::DATA_TYPE_FORM_DATA;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteHostedAppsDataOnExit)) {
+    *remove_mask |= site_data_mask;
+    *origin_mask |= content::BrowsingDataRemover::ORIGIN_TYPE_PROTECTED_WEB;
+  }
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteMediaLicensesOnExit))
+    *remove_mask |= content::BrowsingDataRemover::DATA_TYPE_MEDIA_LICENSES;
+
+  // Note: this will also delete Brave Shields site-specific settings.
```

Comment:
Since this fact may be important for users, maybe consider adding this to the UI?

Reply:
@bsclifton and I discussed getting @tomlowenthal's input on this in the security review, but may as well do it here. One thing to note, though, is that it behaves the same as the setting in the Advanced tab which even now would clear Shields' site-specific settings without warning the user.